keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,266 @@
+- [ ]  (R) Production readiness review completed
+- [ ]  (R) Production readiness review approved
+- [ ]  "Implementation History" section is up-to-date for milestone
+- [ ]  User-facing documentation has been created in [kubernetes/website](https://git.k8s.io/website), for publication to [kubernetes.io](https://kubernetes.io/)
+- [ ]  Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+In Kubernetes, `NodeLifecycleController` applies predefined `NoExecute` taints (e.g., `Unreachable`, `NotReady`) when the nodes are determined to be unhealthy. After the nodes get tainted, `TaintManager` does its due diligence to start deleting running pods on those nodes based on `NoExecute` taints, which can be added by anyone.
+
+In this KEP, we propose to decouple `TaintManager` that performs taint-based pod eviction from `NodeLifecycleController` and make them two separate controllers: `NodeLifecycleController` to add taints to unhealthy nodes and `TaintManager` to perform pod deletion on nodes tainted with NoExecute effect.
+
+This separation not only improves code organization but also makes it easier to improve `TaintManager` or build custom `TaintManager`.
+
+## Motivation
+`NodeLifecycleController` combines two independent functions: 
+   * Adding a pre-defined set of `NoExecute` taints to nodes based on the node conditions 
+   * Performing pod eviction on `NoExecute` taints
+
+Splitting the `NodeLifecycleController` based on the above functionalities will help to disentangle code and make future extensions to either component manageable.
+
+### Goals
+
+* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to enhance separation of concerns, maintainability. This separation also enables cluster-admins to build custom `TaintManager`s and use them in their clusters.
+
+### Non-Goals
+
+The main focus of the KEP is on the separation of concerns between `NodeLifecycleController` and `TaintManager`, which will improve code organization and enable future extensions to `TaintManager`. While it is true that the separation allows cluster operators to build custom `TaintManager` and disable the default `TaintManager`, this is a side-effect of the change rather than the main goal. the KEP emphasizes the benefits of the separation of concerns and briefly mention the option to disable the default TaintManager as a potential side-effect.
+
+* Extend and enhance the current built-in APIs for customization of node taint-based eviction, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.
+* Actual extension and enhancement  of the current `TaintManager`.
```

Comment:
Would it be better to have them as part of the Goals instead if we add "Do Not" ?

- Do not introduce extra maintenance burden or regression 
- Do not introduce incompatible behavior to NodeLifecycleController or TaintManager

Reply:
Nope, the bullets are under `Non-Goals`, so use a positive tone is correct.

Reply:
Are you suggesting we put the two bullets under `Non-Goals`? It sounds weird to me. If they are Non-Goals, would it meany we don't care if the changes will introduce extra maintenance burden and incompatible behavior? Also, these requirements should be self-evident (needless to say) for most changes if not all. 

```
Introduce extra maintenance burden or regression after the code reorganization
Introduce incompatible behavior of NodeLifecycleController or TaintManager
```

Reply:
> Are you suggesting we put the two bullets under Non-Goals?

Yes.

> Also, these requirements should be self-evident (needless to say) for most changes if not all.

It can be debating. But IMO, code restructuring  in this KEP is non-trivial work, and hence may cause regression/incompatibility issue, so listing them as Non-Goals doesn't sound redundant.